Fix unit tests on Windows + enable on bots

DanTup opened the pull request:
(Re-do of https://github.com/sourcegraph/cody/pull/1150 because all the bots won't run on forks)

I don't know how sound all of these changes are, but they fix the failing tests I have on Windows.

1. Use platform-correct paths (drive letters + backslashes on Windows) in tests and debug printing
  Previously lots of code assumed posix-style paths then had mismatches on windows because of slash direction
2. Replace fake Uri class with vscode-uri
  As part of doing (1), I found that `vscode.Uri` was missing. While adding it, I realised there was a mock `Uri` implementation (presumably to avoid importing `vscode` where it won't be available), however the `vscode-uri` package (which does not depend on VS Code) was also already imported, so I removed the fake `Uri` class and just used this. This required updating a bunch of places (`Uri` => `URI`) and I don't know if there are other implications of doing this, but I think it's better.
3. Add launch config to run current test file under debugger
  Makes it easy to debug tests (or run without debugging for the current file) without having to use the terminal
4. Use prettier when editing JSON files like `.vscode/settings.json` and `.vscode/launch.json` (these already seemed to be formatted using Prettier, but it wasn't set explicitly so the formatting gets messed up if you didn't have Prettier as default formatter)

## Test plan

Tested using `pnpm test:unit`.

DanTup commented:
Related question: should the bots be running these tests on Windows too?

olafurpg commented:
Quick comment on vscode-uri, the custom Uri implementation is intentional because vscode-uri is not a comprehensive implementation of vscode.Uri. I'm on the phone right now so can't share concrete details but replacing Uri with URI will break the JetBrains plugin. We have tests to reproduce the cases but they don't run in CI yet because they depend on a Sourcegraph instance.

DanTup commented:
> replacing Uri with URI will break the JetBrains plugin

Ah, thanks for the note! If you have more info on this, it'd be good to add a comment to the class about this for reference.

I reverted that part of the change and hit some issues because the graph tests end up with a mix of shim and real URIs, and there are various comparisons on `fsPath`. The shimmed `fsPath` is incorrect on Windows because it just returns the URI path:

https://github.com/sourcegraph/cody/blob/57dd0b0c5d971ceeee12963a9a6f50ad73357a37/vscode/src/testutils/mocks.ts#L53

I was able to work around this for now by tweaking comparing `URI.toString()` in the failing test (I did first try changing the whole test to use the shim URIs, but that causes other failures because some other code is using vscode-uri and their paths are compared in expectations).

It would probably be good to:

1. Fix fsPath to work correctly in the shim
  I'm not sure if the implications of changing this - if the shim is used in production code, it's currently giving incorrect `fsPath`s on Windows. This should be fixed, but will it change/break anything that may have had existing assumptions about that wrong format?
2. Make it clearer where real/mock URIs should be used
  It's not clear to me whether we should use the shim URI _everywhere_ and never use `vscode-uri`, or if there are reasons to use `vscode-uri` in some places (there are a fair few files using vscode-uri).

I guess an ideal solution would be to remove the shim and find a solution to for the JetBrains plugin - but I don't know how feasible that is (because I'm not familiar with specifics of the problem). If it's not feasible (I suspect if it was easy you would have done it!), I wonder if we could reproduce those issues in tests so the bots can ensure that doesn't get broken by this kind of change 🙃

olafurpg commented:
> I'm not sure if the implications of changing this - if the shim is used in production code, it's currently giving incorrect fsPaths on Windows

This is very insightful, thank you for sharing! We are using the shim for the JetBrains plugin and this could be the root caused behind why autocomplete is not working on Windows.

> Make it clearer where real/mock URIs should be used

Great point! The shim should never be used directly. It's fine to replace all usages of `vscode_shim.Uri` with `vscode.Uri`. In lib/shared, it's fine to use vscode-uri. I'll follow up with a PR to improve documentation.

The only reason we need the shim to be 100% compatible with `vscode.Uri` is because we replace all imports on the `vscode` module with the shim when running the agent (which we use from JetBrains and Neovim). Originally, I tried using vscode-uri `URI` in the shim but it crashed at runtime because vscode-uri `URI` is not a drop-in replacement of `vscode.Uri`.

philipp-spiess reviewed (approved):
Sweet! Let's enable unit tests on windows on CI too! We do this for integration tests already, you just need to define a matrix for the GH workflow like this: https://github.com/sourcegraph/cody/blob/main/.github/workflows/ci.yml#L59-L60

DanTup commented on vscode/src/completions/context/graph-section-observer.test.ts:
```diff
@@ -116,7 +116,7 @@ describe('GraphSectionObserver', () => {
     it('loads visible documents when it loads', () => {
         expect(sectionObserver.debugPrint()).toMatchInlineSnapshot(`
-          "file:/document1.ts
+          "${document1Uri}
```
> we won't be able to use `pnpm test -u` to update the snapshots again

Oh, I didn't know about this. I'll take a look.

I don't seem to be able to mark this PR back as draft, but I want to take another look at it now I have some more context about the vscode-uri issues (for ex. `Uri.join`).

DanTup replied:
For this particular cases, I think we can change `debugPrint()` to use `path.posix.normalize` instead of `path.normalize` and the debug output would always be forward slashes.

However, there are some other tests further down where the values being verified are not from a `debugPrint()` call and look like non-debug values (for example the output of `getContextAtPosition`). I think to preserve the ability to update snapshots, we'd need to wrap the calls to functions like `getContextAtPosition` with something that re-writes the relative paths to use forward slashes? (I'm assuming it's not reasonable to change the slash direction in the underlying data structures because they might be used in ways that need to match the current platform?).

DanTup replied:
I was unable to get `pnpm test:unit -u` to actually update anything (it just reports failures when the snapshots are wrong, like without `-u`). I'm not sure why, but I changed how this works a little to avoid this change. See below...

philipp-spiess replied:
> I was unable to get pnpm test:unit -u to actually update anything (it just reports failures when the snapshots are wrong, like without -u). I'm not sure why, but I changed how this works a little to avoid this change. See below...

Huh interesting. Maybe pnpm did not forward the update flag correctly? I do think I run `pnpm test` from the repo root instead of `pnpm test:unit` from the vscode directory but that shouldn't be making the difference? If the test failes and the UI looks like this you can also hit `u` to update: 

<img width="1353" alt="Screenshot 2023-10-09 at 10 50 46" src="https://github.com/sourcegraph/cody/assets/458591/cf4bb9f0-5a50-4976-aafb-85f10fb891c2">

DanTup replied:
Weird - I can't get `-u` to work anywhere (with `pnpm test` the flag is unsupported, and it seems like `pnpm test:unit` passes flags through, but it just doesn't seem to work). Running `pnpm test` so it runs interactively and then pressing `u` at the end _does_ work though.

I didn't break every test, but I broke a few and used this, and it generated the same expected text (even on Windows), so I think this concern is addressed. Although I'll note there's still a little bit of wonkiness here because we're calling `vscode.workspace.asRelativePath` with paths that are not inside the workspace in tests, so we get back the URI string (not a file path) and then we're sending that into `path.posix.normalize` (which is really intended for paths, not string URIs).

I think it would probably be better to only pass paths (eg. `.fsPath`) to `vscode.workspace.asRelativePath` (which will remove the `file:` prefix here) although if there's any intention to ultimately support non-file URIs (such as virtual workspaces, Live Share, etc.) that might be the wrong direction and instead we should start using URIs everywhere and not trying to make relative paths.

philipp-spiess commented:
Regarding tests on windows: All you need is to add windows here #1315 🙂

DanTup commented:
I pushed some changes, though the changes to each set of tests is a little different:

- debugPrint() now always uses posix paths, that way it can easily compare strings written in that form (and my assumption is that this is just a debug string and using `/` on Windows will cause no issues)
- graph-section-observer tests have the results of `sectionObserver.getContextAtPosition` run through a new function `withPosixPaths` which rewrites the paths so that they can be compared to inline snapshots using posix paths
- graph tests use file URIs built with the native path format rather than hard-coded posix string paths. This is the ideal case because on Windows it's testing everything like-for-like. It works here because there aren't inline snapshots/strings that make it difficult to use the native path format.

DanTup commented:
Seems like there are some other issues running the bots on Windows (and macOS with node 16). Mac looks like connectivity issues downloading wasm libraries, and Windows looks like maybe timeouts and being terminated.

I've reverted that change from this PR to focus on just having the tests pass locally, and will dig into that in another PR (to avoid keeping this open and potentially causing conflicts as people work on other tests).

I also discovered that on the bots Git was using autocrlf which breaks snapshot tests (which combine a mix of a file from disk and some string-build code) which differed from my machine, so I've made this explicitly use lf to make this consistent. I think ideally `auto` should be used (so on Windows the tests are using crlf), but that'll need some changes to the snapshot annotation code (it's also not clear to me whether this difference would affect anything outside of tests so it might not be worthwhile).

DanTup commented:
The Windows issue was a timeout. I increased the timeout to 10 mins (from 5), and they went green in a little over 5 mins.

Slowest steps are:

pnpm install - 2:43
pnpm build - 0:35
pnpm run test:unit --run - 1:12
pnpm store - 0:42

It's possible that part of the issue is that a lack of cache (because it's never succeeded before) is slowing it down a little, but since it's never guaranteed the cache is available, the timeout should probably allow for this. So I've included the change to increase from 5->10 mins here and re-enabled on Windows.

For macOS, the issue is a bit weirder.. on the Node 16 bot, it fails with what seems like a connectivity issue downloading wasm files (and this seems to be 100% consistent):

```
vscode postinstall$ pnpm download-wasm
vscode postinstall: > cody-ai@0.14.0 download-wasm /Users/runner/work/cody/cody/vscode
vscode postinstall: > ts-node-transpile-only ./scripts/download-wasm-modules.ts
vscode postinstall: We are missing 12 files.
vscode postinstall: Error: read ECONNRESET
vscode postinstall:     at TLSWrap.onStreamRead (node:internal/stream_base_commons:217:20) {
vscode postinstall:   errno: -54,
vscode postinstall:   code: 'ECONNRESET',
vscode postinstall:   syscall: 'read'
vscode postinstall: }
vscode postinstall:  ELIFECYCLE  Command failed with exit code 1.
vscode postinstall: Failed
 ELIFECYCLE  Command failed with exit code 1.
```

However I can't understand why this would only affect Node 16 (the Node 20 bot is always green). Maybe there's some difference in Node 16/20 that affects networking on a Mac? Seems odd though. I guess my first questions would be:

1. Why was Node 16 picked?
2. Would it be reasonable to skip the Node 16 both only for Mac?

philipp-spiess commented:
> Why was Node 16 picked?

That was added because VS Code as of (I think) ~ 1month ago used Node 16 in the runtime and since some of the remote desktop setups for VS Code often lag a few versions behind, we got a lot of bug reports after adding code that did not work on Node 16. 

> Would it be reasonable to skip the Node 16 both only for Mac?

Yeah I think so! The main intention with Node 16 is to find unavailable APIs, We only need to run one test for this on one platform 👍 

Excellent work here, thank you 🙇

DanTup commented:
I've tweaked this so only Linux runs on Node 16, and the others on 20. I also rebased on latest because some tests were changed and moved around so there were conflicts, so I resolved those and updated the tests accordingly.

Please take a look :-)

philipp-spiess reviewed (approved):
This approach looks very reasonable. I also love the comments and insights you've left in the code. 


![You Are The Best Thumbs Up](https://media1.giphy.com/media/XE1cLqA94WFTOWqfn1/giphy-downsized-medium.gif?cid=47028fa81rda1t2f7b12e0cyh0rlj4hnehstygaulemz4h55&ep=v1_gifs&rid=giphy-downsized-medium.gif&ct=g)
